File: backend/src/main/java/com/back/ourlog/domain/follow/service/FollowService.java
package com.back.ourlog.domain.follow.service;

import com.back.ourlog.domain.follow.dto.FollowUserResponse;
import com.back.ourlog.domain.follow.entity.Follow;
import com.back.ourlog.domain.follow.enums.FollowStatus;
import com.back.ourlog.domain.follow.repository.FollowRepository;
import com.back.ourlog.domain.user.entity.User;
import com.back.ourlog.domain.user.repository.UserRepository;
import com.back.ourlog.global.exception.CustomException;
import com.back.ourlog.global.exception.ErrorCode;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class FollowService {

    private final FollowRepository followRepository;
    private final UserRepository userRepository;

    @PersistenceContext
    private EntityManager em;

    // 팔로우..
    @Transactional
    public void follow(Integer followerId, Integer followeeId) {
        if (followerId.equals(followeeId)) {
            throw new CustomException(ErrorCode.CANNOT_FOLLOW_SELF);
        }

        // 기존 Follow 있으면 처리..
        List<Follow> existingList = followRepository.findAllByFollowerIdAndFolloweeId(followerId, followeeId);
        for (Follow f : existingList) {
            if (f.getStatus() == FollowStatus.ACCEPTED || f.getStatus() == FollowStatus.PENDING) {
                throw new CustomException(ErrorCode.FOLLOW_ALREADY_EXISTS);
            } else {
                followRepository.delete(f); // 이전 REJECTED 같은 건 삭제..
            }
        }

        // 역방향 PENDING 상태가 있다면 자동 수락..
        Optional<Follow> reversePending = followRepository.findByFollowerIdAndFolloweeIdAndStatus(
                followeeId, followerId, FollowStatus.PENDING);

        if (reversePending.isPresent()) {
            Follow reverse = reversePending.get();
            reverse.accept();
            reverse.getFollower().increaseFollowingsCount();
            reverse.getFollowee().increaseFollowersCount();
            followRepository.save(reverse);

            // ✅ forward 관계도 추가 (나 → 상대방)
            User follower = userRepository.findById(followerId)
                    .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));
            User followee = userRepository.findById(followeeId)
                    .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));

            Follow forward = new Follow(follower, followee);
            forward.accept();
            followRepository.save(forward);

            follower.increaseFollowingsCount();
            followee.increaseFollowersCount();

            return;
        }



        // 역방향이 ACCEPTED인 경우 → 쌍방으로 만들어야 함..
        Optional<Follow> reverseAccepted = followRepository.findByFollowerIdAndFolloweeIdAndStatus(
                followeeId, followerId, FollowStatus.ACCEPTED);

        if (reverseAccepted.isPresent()) {
            User follower = userRepository.findById(followerId)
                    .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));
            User followee = userRepository.findById(followeeId)
                    .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));

            Follow follow = new Follow(follower, followee);
            follow.accept();
            followRepository.save(follow);

            follower.increaseFollowingsCount();
            followee.increaseFollowersCount();
            return;
        }

        // 아무 관계도 없으면 새로 PENDING 생성..
        User follower = userRepository.findById(followerId)
                .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));
        User followee = userRepository.findById(followeeId)
                .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));

        Follow follow = new Follow(follower, followee); // 상태는 기본 PENDING..
        followRepository.save(follow);

        follower.increaseFollowingsCount();
        followee.increaseFollowersCount();
    }


    // 언팔로우..
    @Transactional
    public void unfollow(Integer myUserId, Integer otherUserId) {
        List<Follow> follows = followRepository.findAllByUsersEitherDirection(myUserId, otherUserId);

        if (follows.isEmpty()) {
            throw new CustomException(ErrorCode.FOLLOW_NOT_FOUND);
        }

        for (Follow follow : follows) {
            followRepository.delete(follow);
            follow.getFollower().decreaseFollowingsCount();
            follow.getFollowee().decreaseFollowersCount();
        }

        followRepository.flush();
        em.clear();
    }

    // 내가 팔로우한 유저 목록 조회..
    public List<FollowUserResponse> getFollowings(Integer userId) {
        userRepository.findById(userId)
                .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));

        List<Follow> followings = followRepository.findFollowingsByUserId(userId);

        return followings.stream()
                .map(f -> FollowUserResponse.fromFollowee(f, true))
                .toList();
    }


    // 나를 팔로우한 유저 목록 조회..
    public List<FollowUserResponse> getFollowers(Integer userId) {
        userRepository.findById(userId)
                .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));

        List<Follow> follows = followRepository.findFollowersByUserId(userId);

        return follows.stream()
                .map(f -> {
                    Integer followerId = f.getFollower().getId(); // 여기에 선언..
                    boolean isFollowing = followRepository
                            .findByFollowerIdAndFolloweeIdAndStatus(userId, followerId, FollowStatus.ACCEPTED)
                            .isPresent();

                    System.out.println("[FOLLOWERS DEBUG] userId = " + userId +
                            ", followerId = " + followerId +
                            ", isFollowing = " + isFollowing);

                    return FollowUserResponse.fromFollower(f, isFollowing);
                })
                .toList();
    }


    // 팔로우 요청을 수락 상태로 변경..
    @Transactional
    public void acceptFollow(Integer followId) {
        Follow follow = followRepository.findById(followId)
                .orElseThrow(() -> new CustomException(ErrorCode.FOLLOW_NOT_FOUND));

        if (follow.getStatus() == FollowStatus.ACCEPTED) {
            throw new CustomException(ErrorCode.FOLLOW_ALREADY_ACCEPTED);
        }

        if (follow.getStatus() == FollowStatus.REJECTED) {
            throw new CustomException(ErrorCode.FOLLOW_ALREADY_REJECTED);
        }

        follow.accept();

        followRepository.save(follow);
        followRepository.flush();

    }

    // 팔로우 요청을 거절 상태로 변경..
    @Transactional
    public void rejectFollow(Integer followId) {
        Follow follow = followRepository.findById(followId)
                .orElseThrow(() -> new CustomException(ErrorCode.FOLLOW_NOT_FOUND));

        if (follow.getStatus() == FollowStatus.REJECTED) {
            throw new CustomException(ErrorCode.FOLLOW_ALREADY_REJECTED);
        }

        if (follow.getStatus() == FollowStatus.ACCEPTED) {
            throw new CustomException(ErrorCode.FOLLOW_ALREADY_ACCEPTED);
        }

        follow.reject();
    }

    // 내가 보낸 팔로우 요청 목록 (PENDING 상태)..
    public List<FollowUserResponse> getSentPendingRequests(Integer userId) {
        userRepository.findById(userId)
                .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));

        List<Follow> sentRequests = followRepository.findSentPendingRequestsByUserId(userId);

        return sentRequests.stream()
                .map(FollowUserResponse::fromFollowee)
                .toList();
    }

    // 내가 받은 팔로우 요청 목록 (PENDING 상태)..
    public List<FollowUserResponse> getPendingRequests(Integer userId) {
        userRepository.findById(userId)
                .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));

        List<Follow> pendingFollows = followRepository.findPendingRequestsByUserId(userId);

        return pendingFollows.stream()
                .map(FollowUserResponse::fromFollower)
                .toList();
    }
}

File: backend/src/main/java/com/back/ourlog/domain/follow/service/FollowService.kt
package com.back.ourlog.domain.follow.service

import com.back.ourlog.domain.follow.dto.FollowUserResponse
import com.back.ourlog.domain.follow.entity.Follow
import com.back.ourlog.domain.follow.enums.FollowStatus
import com.back.ourlog.domain.follow.repository.FollowRepository
import com.back.ourlog.domain.user.entity.User
import com.back.ourlog.domain.user.repository.UserRepository
import com.back.ourlog.global.exception.CustomException
import com.back.ourlog.global.exception.ErrorCode
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class FollowService(
    private val followRepository: FollowRepository,
    private val userRepository: UserRepository,
) {
    // 팔로우 요청을 처리하는 함수
    @Transactional
    fun follow(followerId: Int, followeeId: Int) {
        // require 함수를 사용해 자기 자신을 팔로우하는 요청을 막고, 예외를 발생시킵니다.
        require(followerId != followeeId) { throw CustomException(ErrorCode.CANNOT_FOLLOW_SELF) }

        // 기존에 팔로우 관계가 있는지 두 가지 경우(정방향, 역방향)로 데이터베이스를 조회합니다.
        val existingFollow = followRepository.findByFollowerIdAndFolloweeId(followerId, followeeId)
        val reversedFollow = followRepository.findByFollowerIdAndFolloweeId(followeeId, followerId)

        // when 구문을 사용해 다양한 상황(이미 팔로우, 맞팔, 신규 요청)을 깔끔하게 처리합니다.
        when {
            // Case 1: 이미 팔로우 중이거나 요청을 보낸 경우
            existingFollow?.status == FollowStatus.ACCEPTED || existingFollow?.status == FollowStatus.PENDING -> {
                throw CustomException(ErrorCode.FOLLOW_ALREADY_EXISTS)
            }

            // Case 2: 상대방이 나에게 보낸 'PENDING' 요청이 있는 경우 -> 맞팔 수락 로직 제거 및 예외 발생
            reversedFollow?.status == FollowStatus.PENDING -> {
                // 즉시 수락하지 않고, 예외를 발생시켜 클라이언트가 '요청 수락' 버튼을 누르도록 유도
                throw CustomException(ErrorCode.FOLLOW_REQUEST_EXISTS)
            }

            // Case 3: 기존 관계가 없거나 'REJECTED' 상태, 상대가 나를 팔로우 중인 경우 모두 포함
            // REJECTED 관계가 있으면 삭제하고, 새로운 PENDING 팔로우 요청을 생성합니다.
            else -> {
                existingFollow?.let { followRepository.delete(it) } // 내가 보냈던 REJECTED 요청 등이 있다면 삭제
                val newFollow = Follow(findUserById(followerId), findUserById(followeeId))
                followRepository.save(newFollow)
                updateFollowCounts(newFollow)
            }
        }
    }

    // 언팔로우 요청을 처리하는 함수
    @Transactional
    fun unfollow(myUserId: Int, otherUserId: Int) {
        // 1. myUserId가 otherUserId를 팔로우하는 '단방향' 관계만 정확히 찾는다
        val followToDelete = followRepository.findByFollowerIdAndFolloweeId(myUserId, otherUserId)
            ?: throw CustomException(ErrorCode.FOLLOW_NOT_FOUND)

        // 2. 해당 관계 '하나'의 팔로워/팔로이 카운트만 감소시킨다
        followToDelete.follower.decreaseFollowingsCount() // 나의 팔로잉 수 감소
        followToDelete.followee.decreaseFollowersCount() // 상대방의 팔로워 수 감소

        // 3. 데이터베이스에서 해당 팔로우 관계 '하나만' 삭제한다
        followRepository.delete(followToDelete)
    }

    // 내가 팔로우한 유저 목록을 조회하는 함수
    fun getFollowings(userId: Int): List<FollowUserResponse> {
        val user = findUserById(userId)
        // findFollowingsByFollowerIdAndStatus 메서드를 호출해 ACCEPTED 상태인 팔로우 목록을 가져옵니다.
        // map 함수를 사용해 각 Follow 객체를 FollowUserResponse DTO로 변환합니다.
        return followRepository.findFollowingsByFollowerIdAndStatus(user.id!!, FollowStatus.ACCEPTED).map {
            FollowUserResponse.fromFollowee(it, true)
        }
    }

    // 나를 팔로우한 유저 목록을 조회하는 함수
    fun getFollowers(userId: Int): List<FollowUserResponse> {
        val user = findUserById(userId)
        // 나를 팔로우하는 유저 목록을 가져옵니다.
        val followers = followRepository.findFollowersByFolloweeIdAndStatus(user.id!!, FollowStatus.ACCEPTED)

        // 내가 팔로우하는 모든 사람의 ID를 Set으로 미리 가져와 메모리에 저장합니다.
        // 이렇게 하면 데이터베이스를 여러 번 조회하지 않아도 됩니다.
        val myFollowingIds = followRepository.findFollowingsByFollowerIdAndStatus(user.id!!, FollowStatus.ACCEPTED).map { it.followee.id }.toSet()

        // map 함수를 사용해 각 팔로워에 대해 내가 맞팔로우하는지 여부를 확인하고 DTO로 변환합니다.
        return followers.map { follow ->
            val isFollowingBack = myFollowingIds.contains(follow.follower.id!!)
            FollowUserResponse.fromFollower(follow, isFollowingBack)
        }
    }

    // 팔로우 요청을 수락하는 함수
    @Transactional
    fun acceptFollow(followId: Int) {
        val follow = findFollowById(followId)
        // when 구문을 사용해 현재 팔로우 상태에 따라 다른 동작을 처리합니다.
        when (follow.status) {
            FollowStatus.ACCEPTED -> throw CustomException(ErrorCode.FOLLOW_ALREADY_ACCEPTED)
            FollowStatus.REJECTED -> throw CustomException(ErrorCode.FOLLOW_ALREADY_REJECTED)
            FollowStatus.PENDING -> {
                // PENDING 상태일 때만 수락 처리합니다.
                follow.accept()
                updateFollowCounts(follow)
            }
        }
    }

    // 팔로우 요청을 거절하는 함수
    @Transactional
    fun rejectFollow(followId: Int) {
        val follow = findFollowById(followId)
        // when 구문을 사용해 현재 팔로우 상태에 따라 다른 동작을 처리합니다.
        when (follow.status) {
            FollowStatus.REJECTED -> throw CustomException(ErrorCode.FOLLOW_ALREADY_REJECTED)
            FollowStatus.ACCEPTED -> throw CustomException(ErrorCode.FOLLOW_ALREADY_ACCEPTED)
            FollowStatus.PENDING -> follow.reject()
        }
    }

    // 팔로워와 팔로이의 팔로우/팔로잉 카운트를 증가시킵니다.
    private fun updateFollowCounts(vararg follows: Follow) {
        follows.forEach { follow ->
            follow.follower.increaseFollowingsCount()
            follow.followee.increaseFollowersCount()
        }
    }

    // 내가 보낸 팔로우 요청 목록을 조회하는 함수
    fun getSentPendingRequests(userId: Int): List<FollowUserResponse> {
        val user = findUserById(userId)
        return followRepository.findSentPendingRequestsByFollowerIdAndStatus(user.id!!, FollowStatus.PENDING)
            .map(FollowUserResponse::fromFollowee)
    }

    // 내가 받은 팔로우 요청 목록을 조회하는 함수
    fun getPendingRequests(userId: Int): List<FollowUserResponse> {
        val user = findUserById(userId)
        return followRepository.findPendingRequestsByFolloweeIdAndStatus(user.id!!, FollowStatus.PENDING)
            .map(FollowUserResponse::fromFollower)
    }

    // 주어진 ID로 유저를 찾아 반환하는 함수
    private fun findUserById(userId: Int): User {
        return userRepository.findByIdOrNull(userId)
            ?: throw CustomException(ErrorCode.USER_NOT_FOUND)
    }

    // 주어진 ID로 팔로우 관계를 찾아 반환하는 함수
    private fun findFollowById(followId: Int): Follow {
        return followRepository.findByIdOrNull(followId)
            ?: throw CustomException(ErrorCode.FOLLOW_NOT_FOUND)
    }
}
